lib/rpc/client/client.go

```diff
@@ -226,16 +227,81 @@ func (r *client) Close() error {
 // DialRPC returns RPC client for the provided addr.
 type DialRPC func(ctx context.Context, addr string) (Client, error)
 
+type ClientCache struct {
+	sync.RWMutex
+	clients map[string]Client
+}
+
 // DefaultDialRPC is the default RPC client factory function.
 // It creates a new client based on address details.
-func DefaultDialRPC(caFile, certFile, keyFile string) DialRPC {
+func (c *ClientCache) DefaultDialRPC(caFile, certFile, keyFile string) DialRPC {
 	return func(ctx context.Context, addr string) (Client, error) {
+		client := c.getClientFromCache(addr)
+		if client != nil {
+			return client, nil
+		}
+
 		config := Config{
 			Address:  fmt.Sprintf("%s:%d", addr, rpc.Port),
 			CAFile:   caFile,
 			CertFile: certFile,
 			KeyFile:  keyFile,
 		}
-		return NewClient(ctx, config)
+		client, err := NewClient(context.Background(), config)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		c.Lock()
+		if c.clients == nil {
+			c.clients = make(map[string]Client)
+		}
+		c.clients[addr] = client
+		c.Unlock()
+
+		return client, nil
+	}
+}
+
+func (c *ClientCache) getClientFromCache(addr string) Client {
+	c.RLock()
+	defer c.RUnlock()
+
+	if c.clients == nil {
+		return nil
+	}
+
+	if client, ok := c.clients[addr]; ok {
+		return client
+	}
+
+	return nil
+}
+
+// CloseMissingMembers will removed gRPC clients from the cache that don't appear to be part
+// of the cluster. It will do so after sleeping for the provided timeout.
+func (c *ClientCache) CloseMissingMembers(currentMembers []*pb.MemberStatus) {
+	currentMap := make(map[string]interface{})
+
+	for _, member := range currentMembers {
+		currentMap[member.Addr] = nil
+	}
+
+	removed := make(map[string]Client)
+
+	c.Lock()
+	for addr, client := range c.clients {
+		if _, ok := currentMap[addr]; !ok {
+			removed[addr] = client
+		}
+	}
+
+	for addr := range removed {
+		delete(c.clients, addr)
+	}
+	c.Unlock()
+
+	for _, client := range removed {
+		client.Close()
```

Comment:
It looks like the only scenario for an error to be returned is if someone tries to close the client multiple times, but it's possible this might change in a client update, so I added a debug log but am not sure we should do anything above that.